build_modules/lib/src/meta_module_clean_builder.dart

```diff
@@ -0,0 +1,125 @@
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:collection';
+import 'dart:convert';
+
+import 'package:graphs/graphs.dart';
+
+import 'package:build/build.dart';
+import 'meta_module.dart';
+import 'meta_module_builder.dart';
+import 'modules.dart';
+
+/// The extension for serialized clean meta module assets.
+const metaModuleCleanExtension = '.meta_module_clean';
+
+/// Map of [AssetId] to corresponding primary [AssetId] within the same
+/// [Module].
+final _assetToModule =
+    new Resource<Map<AssetId, Module>>(() => {}, dispose: (map) => map.clear());
+
+/// Returns a set of all modules transitvely reachable from the provided meta
+/// module asset.
+Future<Set<Module>> _transitiveModules(BuildStep buildStep, AssetId metaAsset,
+    Map<AssetId, Module> assetToModule) async {
+  var dependentModules = new Set<Module>();
+  // Ensures we only process a meta file once.
+  var seenMetas = new Set<AssetId>()..add(metaAsset);
+  var meta = new MetaModule.fromJson(
+      json.decode(await buildStep.readAsString(buildStep.inputId))
+          as Map<String, dynamic>);
+  var nextModules = <Module>[];
+  nextModules.addAll(meta.modules);
+  while (nextModules.isNotEmpty) {
+    var module = nextModules.removeLast();
+    dependentModules.add(module);
+    for (var source in module.sources) {
+      assetToModule[source] = module;
+    }
+    for (var dep in module.directDependencies) {
+      var depMetaAsset = new AssetId(dep.package, 'lib/$metaModuleExtension');
+      if (!seenMetas.contains(depMetaAsset)) {
+        seenMetas.add(depMetaAsset);
+        if (await buildStep.canRead(depMetaAsset)) {
+          var depMeta = new MetaModule.fromJson(
+              json.decode(await buildStep.readAsString(depMetaAsset))
+                  as Map<String, dynamic>);
+          nextModules.addAll(depMeta.modules);
+        } else {
+          log.info('Unable to find meta module for dependency: $dep '
+              'Are you missing a dependency on package:${dep.package}?');
+        }
+      }
+    }
+  }
+  return dependentModules;
+}
+
+/// Merges the modules in a strongly connected component.
+///
+/// Note this will clean the module dependencies as the merge happens.
+/// The result will be that all dependencies are primary sources.
+Module _mergeComponent(
+    List<Module> connectedComponent, Map<AssetId, Module> assetToModule) {
+  var sources = new Set<AssetId>();
+  var deps = new Set<AssetId>();
+  // Sort the modules to deterministicly select the primary source.
+  var components = new SplayTreeSet<Module>(
+      (a, b) => a.primarySource.compareTo(b.primarySource))
+    ..addAll(connectedComponent);
+  var primarySource = components.first.primarySource;
+  for (var module in connectedComponent) {
+    sources.addAll(module.sources);
+    for (var dep in module.directDependencies) {
+      var depModule = assetToModule[dep];
```

Comment:
It is a bit hard to follow, but will this map contain un-merged modules? We need to make sure it ends up pointing at the merged modules returned by this function, right?